### PR TITLE
add OPTEE_RUST_ENABLE option to common.mk

### PR DESCRIPTION
**Pull request by fanrong1992**
Signed-off-by: Rong Fan <fanrong03@baidu.com>
Yuan Zhuang <zhuangyuan04@baidu.com>

### CODE REVIEW

File: br-ext/package/optee_rust_ext/Config.in [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+config BR2_PACKAGE_OPTEE_RUST_EXT_SITE
+	string "OP-TEE rust path"
+	default ""
+	help
+	  The path to the source directory of OP-TEE Rust SDK
+
+if BR2_PACKAGE_OPTEE_RUST_EXT
+
+config BR2_PACKAGE_OPTEE_RUST_EXT_CROSS_COMPILE
+	string "OPTEE cross compiler"
```
Comment (jbech-linaro) [inline br-ext/package/optee_rust_ext/Config.in:15]: s/OPTEE/OP-TEE/

File: br-ext/package/optee_rust_ext/optee_rust_ext.mk [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+define OPTEE_RUST_EXT_INSTALL_TARGET_CMDS
+	@$(foreach f,$(wildcard $(@D)/examples/*/ta/target/$(TA_TARGET)/release/*.ta), \
+		mkdir -p $(TARGET_DIR)/lib/optee_armtz && \
+		echo Installing $f && \
+		$(INSTALL) -v -p --mode=444 \
+			--target-directory=$(TARGET_DIR)/lib/optee_armtz $f \
+			&&) true
+	@$(foreach f,$(wildcard $(@D)/examples/*/host/target/$(HOST_TARGET)/release/*-rs), \
+		echo Installing $f && \
+		$(INSTALL) -v -p --target-directory=$(TARGET_DIR)/usr/bin $f \
```
Comment (jbech-linaro) [inline br-ext/package/optee_rust_ext/optee_rust_ext.mk:37]: Here we're putting the Rust TA's in `/usr/bin/` instead of `/lib/optee_armtz/`. What is the motivation behind that? (I guess an upcoming patch in `optee_client` will look for Rust TA's there).

Comment (jenswi-linaro) [reply br-ext/package/optee_rust_ext/optee_rust_ext.mk:37]: Why do Rust TAs need a special place? I certainly hope they are loaded in the same way.

File: br-ext/package/optee_rust_ext/optee_rust_ext.mk [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+export OPTEE_OS_INCLUDE=$(OPTEE_DIR)/optee_os/out/arm/export-ta_arm64/include
+
+define OPTEE_RUST_EXT_BUILD_CMDS
+	@$(foreach f,$(wildcard $(@D)/examples/*/Makefile), \
+		echo Building $f && \
+		$(MAKE) -C $(dir $f) &&) true
+endef
+
+define OPTEE_RUST_EXT_INSTALL_TARGET_CMDS
+	@$(foreach f,$(wildcard $(@D)/examples/*/ta/target/$(TA_TARGET)/release/*.ta), \
```
[the remark this answers is not in the document]
Comment (fanrong1992) [reply br-ext/package/optee_rust_ext/optee_rust_ext.mk:29]: This foreach puts the Rust TA's in `/lib/optee_armtz/`.

File: br-ext/package/optee_rust_ext/optee_rust_ext.mk (hunk 2 above)
Comment (fanrong1992) [reply br-ext/package/optee_rust_ext/optee_rust_ext.mk:37]: This `foreach` puts the Rust CA's in `/usr/bin`.

Comment (jenswi-linaro) [reply br-ext/package/optee_rust_ext/optee_rust_ext.mk:37]: I see, thanks.

File: common.mk [hunk 4]
```diff
@@ -275,6 +276,12 @@ BR2_PACKAGE_OPTEE_EXAMPLES_EXT ?= y
 BR2_PACKAGE_OPTEE_EXAMPLES_EXT_CROSS_COMPILE ?= $(CROSS_COMPILE_S_USER)
 BR2_PACKAGE_OPTEE_EXAMPLES_EXT_SDK ?= $(OPTEE_OS_TA_DEV_KIT_DIR)
 BR2_PACKAGE_OPTEE_EXAMPLES_EXT_SITE ?= $(OPTEE_EXAMPLES_PATH)
+OPTEE_RUST_ENABLE=n
```
Comment (jenswi-linaro) [inline common.mk:279]: Please make this `OPTEE_RUST_ENABLE ?= n`

File: br-ext/Config.in [hunk 5]
```diff
@@ -2,6 +2,7 @@ source "$BR2_EXTERNAL_OPTEE_PATH/package/optee_os_ext/Config.in"
 source "$BR2_EXTERNAL_OPTEE_PATH/package/optee_client_ext/Config.in"
 source "$BR2_EXTERNAL_OPTEE_PATH/package/optee_test_ext/Config.in"
 source "$BR2_EXTERNAL_OPTEE_PATH/package/optee_examples_ext/Config.in"
+source "$BR2_EXTERNAL_OPTEE_PATH/package/optee_rust_ext/Config.in"
```
Comment (jenswi-linaro) [inline br-ext/Config.in:5]: Since this also is an example I believe `optee_rust_examples_ext` should be more accurate.

**Review by jenswi-linaro — COMMENTED** [review]
Please update the commit message to mention that it also defines buildroot packages for OP-TEE Rust examples.

File: br-ext/package/optee_rust_ext/optee_rust_ext.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbech-linaro) [reply br-ext/package/optee_rust_ext/optee_rust_ext.mk:29]: Aha, I've probably misunderstood Mingshen, I thought that the Rust TA's would have the -rs extension. But this makes sense, thanks.

**Comment by jbech-linaro** [comment]
`Acked-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by etienne-lms** [comment]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by ruchi393** [comment]
`Acked-by: Ruchika Gupta <ruchika.gupta@linaro.org>`

**Comment by jforissier** [comment]
Author doesn't match `Signed-off-by`.

**Review by jforissier — CHANGES_REQUESTED** [review]
- About the commit description
  * We need a subject, a blank line, then a description.
  * Missing blank line before the S-o-b.
  * The commit description should explain how to build and how to run the examples.
- This adds build-time dependencies, they need to be mentioned in the commit text too. Perhaps even in the OP-TEE doc? (https://optee.readthedocs.io/en/latest/building/prerequisites.html#prerequisites)
  * cargo (`sudo apt install cargo`)
  * I get a build error: "the `aarch64-unknown-linux-gnu` target may not be installed". This needs to be fixed. Why not use the same toolchain as for building OP-TEE? (the one coming from `toolchain.mk`).

**Comment by mssun** [comment]
> * This adds build-time dependencies, they need to be mentioned in the commit text too. Perhaps even in the OP-TEE doc? (https://optee.readthedocs.io/en/latest/building/prerequisites.html#prerequisites)

For the build-time dependencies, we have some specific requirements on Rust toolchains for cross compilation. Therefore, we provide a setup script (https://github.com/apache/incubator-teaclave-trustzone-sdk/blob/master/setup.sh) to prepare the building environment for the Rust examples. These will be described in the later PR on documentation.

>   * cargo (`sudo apt install cargo`)

We use `rustup` to install cargo, rustc, and related toolchains.

>   * I get a build error: "the `aarch64-unknown-linux-gnu` target may not be installed". This needs to be fixed. Why not use the same toolchain as for building OP-TEE? (the one coming from `toolchain.mk`).

Basically, Rust has its own toolchains and need to be installed explicitly with `rustup` for the cross compilation.

**Comment by jforissier** [comment]
> For the build-time dependencies, we have some specific requirements on Rust toolchains for cross compilation. Therefore, we provide a setup script (https://github.com/apache/incubator-teaclave-trustzone-sdk/blob/master/setup.sh) to prepare the building environment for the Rust examples. 

OK, that is fine

> These will be described in the later PR on documentation.

Good, but I need at least some basic explanations in this commit (in the Makefiles or in the commit description), because in order to give my opinion on this PR I need to test it, and I cannot test if I have no instructions on how to do so. If these instructions are too long to fit in a commit description, then we have a problem. Consider how we can build and run the OP-TEE examples (`make toolchains; make run`, `optee_example_*`) and the whole test suite (`xtest`, `make check`) in QEMU, with or without virtualization (`XEN_BOOT=y`). It only takes a few commands and I certainly hope it is the same for Rust.

**Comment by mssun** [comment]
Oh, I just saw that @fanrong1992 has force pushed a new version (706d135) with some explanations in the commit message. Hope this can help.

**Comment by jforissier** [comment]
@mssun @fanrong1992 thanks.

Author still doesn't match the first `Signed-off-by`. Please use `git config --global user.name "Rong Fan"` then `git commit --amend`.

I followed the instructions but the build failed to complete, see the log below. I first tried on my local machine, then in the Docker image used for the QEMUv8 CI in our Azure script (it is supposed to be able to build the whole project). Can you help me find out what I am missing? I am now trying with your Docker image (teaclave/teaclave-trustzone-sdk-build:0.2.1) in case it is just a matter of dependencies.

```
$ docker run -it --rm jforissier/optee_os_ci:qemuv8_check 
root@229bc2e343f4:/# cd /root && mkdir optee-qemuv8 && cd optee-qemuv8 && repo init -u https://github.com/OP-TEE/manifest.git -m qemu_v8.xml && repo sync -j10
[...]
root@229bc2e343f4:~/optee-qemuv8# (cd .repo/manifests && git fetch origin pull/200/head && git checkout FETCH_HEAD)
[...]
HEAD is now at 841e08c qemu_v8: add Rust SDK entry
root@229bc2e343f4:~/optee-qemuv8# repo sync              
root@229bc2e343f4:~/optee-qemuv8# cd build
root@229bc2e343f4:~/optee-qemuv8/build# git fetch github pull/521/head && git checkout FETCH_HEAD
[...]
HEAD is now at 706d135 add rust examples to OP-TEE
root@229bc2e343f4:~/optee-qemuv8/build# make toolchains
[...]
root@229bc2e343f4:~/optee-qemuv8/build# (cd ../optee_rust && ./setup.sh)
[...]
root@229bc2e343f4:~/optee-qemuv8/build# make -j10 OPTEE_RUST_ENABLE=y CFG_TEE_RAM_VA_SIZE=0x00300000
[...]
>>> optee_rust_examples_ext 1.0 Building
Building /root/optee-qemuv8/out-br/build/optee_rust_examples_ext-1.0/examples/acipher-rs/Makefile
make[4]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[4]: cargo: No such file or directory
root@229bc2e343f4:~/optee-qemuv8/build# sudo apt install -y cargo
root@229bc2e343f4:~/optee-qemuv8/build# make OPTEE_RUST_ENABLE=y CFG_TEE_RAM_VA_SIZE=0x00300000 buildroot
[...]
>>> optee_rust_examples_ext 1.0 Building
Building /root/optee-qemuv8/out-br/build/optee_rust_examples_ext-1.0/examples/acipher-rs/Makefile
make[4]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
  Downloaded proc-macro2 v0.4.30
  Downloaded quote v0.6.13
  Downloaded syn v0.15.44
  Downloaded unicode-xid v0.1.0
  Downloaded libc v0.2.95
  Downloaded uuid v0.8.2
  Downloaded hex v0.3.2
  Downloaded uuid v0.7.2
  Downloaded 8 crates (849.8 KB) in 1.22s
   Compiling proc-macro2 v0.4.30
   Compiling unicode-xid v0.1.0
   Compiling libc v0.2.95
   Compiling syn v0.15.44
   Compiling uuid v0.8.2
   Compiling optee-teec-sys v0.1.0 (/root/optee-qemuv8/out-br/build/optee_rust_examples_ext-1.0/optee-teec/optee-teec-sys)
   Compiling hex v0.3.2
   Compiling uuid v0.7.2
error[E0463]: can't find crate for `std`
  |
  = note: the `aarch64-unknown-linux-gnu` target may not be installed
```

**Comment by fanrong1992** [comment]
> ```
> HEAD is now at 706d135 add rust examples to OP-TEE
> root@229bc2e343f4:~/optee-qemuv8/build# make toolchains
> [...]
> root@229bc2e343f4:~/optee-qemuv8/build# (cd ../optee_rust && ./setup.sh)
> [...]
> root@229bc2e343f4:~/optee-qemuv8/build# make -j10 OPTEE_RUST_ENABLE=y CFG_TEE_RAM_VA_SIZE=0x00300000
> [...]
> ```

`./setup.sh` may not run some of the sub-commands inside, please try to use `source setup.sh`.

**Comment by jforissier** [comment]
> > ```
> > HEAD is now at 706d135 add rust examples to OP-TEE
> > root@229bc2e343f4:~/optee-qemuv8/build# make toolchains
> > [...]
> > root@229bc2e343f4:~/optee-qemuv8/build# (cd ../optee_rust && ./setup.sh)
> > [...]
> > root@229bc2e343f4:~/optee-qemuv8/build# make -j10 OPTEE_RUST_ENABLE=y CFG_TEE_RAM_VA_SIZE=0x00300000
> > [...]
> > ```
> 
> `./setup.sh` may not run some of the sub-commands inside, please try to use `source setup.sh`.

Thanks for the tip. I just did `source $HOME/.cargo/env` before building (`make ...`) and the build completed. Please update the commit description with `source setup.sh` instead of `./setup.sh`. While you're at it, I'd prefer the subject to be "Add Rust examples" (First letter uppercase, and the OP-TEE mention is useless).

Now a couple more issues/questions...

- I got an error the second time I ran `make -j10 OPTEE_RUST_ENABLE=y CFG_TEE_RAM_VA_SIZE=0x00300000 run`:
  ```
  mv: '/home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_rust_examples_ext-1.0/examples/supp_plugin-rs/plugin/target/aarch64-unknown-linux-gnu/release/libsyslog_plugin.so' and '/home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_rust_examples_ext-1.0/examples/supp_plugin-rs/plugin/target/aarch64-unknown-linux-gnu/release/ef620757-fa2b-4f19-a1c4-6e51cfe4c0f9.plugin.so' are the same file
  ```
- It looks like there is a version mismatch somewhere:
  ```
  Welcome to Buildroot, type root or test to login
  buildroot login: test
  $ random-rs -h
  random-rs: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by random-rs)
  ```

**Comment by jenswi-linaro** [comment]
> `./setup.sh` may not run some of the sub-commands inside, please try to use `source setup.sh`.

How does that works with bourne shell, ash, ksh, zsh or even csh and tcsh?

**Comment by jforissier** [comment]
> > `./setup.sh` may not run some of the sub-commands inside, please try to use `source setup.sh`.
> 
> How does that work with bourne shell, ash, ksh, zsh or even csh and tcsh?

I wondered the same, and noticed that ~/.cargo/env also assumes a `/bin/sh`-compatible shell (it uses `export` to update the `PATH`).

**Comment by jenswi-linaro** [comment]
> > > `./setup.sh` may not run some of the sub-commands inside, please try to use `source setup.sh`.
> > 
> > 
> > How does that work with bourne shell, ash, ksh, zsh or even csh and tcsh?
> 
> I wondered the same, and noticed that ~/.cargo/env also assumes a `/bin/sh`-compatible shell (it uses `export` to update the `PATH`).

Does that mean that you need so run a /bin/sh-compatible shell for this to work? Is there a way of wrapping this in the makefiles? We do set the shell to `bash` in the makefiles so there we can rely on bash being used.

**Comment by jforissier** [comment]
> > > > `./setup.sh` may not run some of the sub-commands inside, please try to use `source setup.sh`.
> > > 
> > > 
> > > How does that work with bourne shell, ash, ksh, zsh or even csh and tcsh?
> > 
> > 
> > I wondered the same, and noticed that ~/.cargo/env also assumes a `/bin/sh`-compatible shell (it uses `export` to update the `PATH`).
> 
> Does that mean that you need so run a /bin/sh-compatible shell for this to work? Is there a way of wrapping this in the makefiles? We do set the shell to `bash` in the makefiles so there we can rely on bash being used.

Good idea, perhaps the `buildroot` target should take care of sourcing this env file since the Rust examples are built as a part of this target.

File: common.mk [hunk 6]
```diff
@@ -308,6 +315,9 @@ append-br2-vars = $(foreach var,$(filter BR2_%,$(.VARIABLES)),$(call append-var,
 
 .PHONY: buildroot
 buildroot: optee-os
+ifeq ($(OPTEE_RUST_ENABLE),y)
+    @(cd $(OPTEE_RUST_PATH) && source setup.sh)
```
Comment (jenswi-linaro) [inline common.mk:319]: I believe that the shell is terminated after each line, so this configuration will be rather short lived.

**Comment by fanrong1992** [comment]
> Now a couple more issues/questions...
> 
> * I got an error the second time I ran `make -j10 OPTEE_RUST_ENABLE=y CFG_TEE_RAM_VA_SIZE=0x00300000 run`:
>   ```
>   mv: '/home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_rust_examples_ext-1.0/examples/supp_plugin-rs/plugin/target/aarch64-unknown-linux-gnu/release/libsyslog_plugin.so' and '/home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_rust_examples_ext-1.0/examples/supp_plugin-rs/plugin/target/aarch64-unknown-linux-gnu/release/ef620757-fa2b-4f19-a1c4-6e51cfe4c0f9.plugin.so' are the same file
>   ```

Thank you for your comments. This problem will be resolved in next commit in teaclave-trustzone-sdk.
> * It looks like there is a version mismatch somewhere:
>   ```
>   Welcome to Buildroot, type root or test to login
>   buildroot login: test
>   $ random-rs -h
>   random-rs: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by random-rs)
>   ```

Sorry, we can't reproduce this error. We compile OP-TEE and rust examples on Ubuntu 20.04 with libc 2.31, and `random-rs` runs normally. We also run `random-rs` successfully in docker `jforissier/optee_os_ci:qemuv8_check`. Could you please give more details?

File: common.mk (hunk 6 above)
Comment (fanrong1992) [reply common.mk:319]: Thank you for your comment. That's right, so we put all the environment variables in `br-ext/package/optee_rust_examples_ext/optee_rust_examples_ext.mk` and only some setup and initialization operations in `setup.sh`.

Comment (jenswi-linaro) [reply common.mk:319]: Why is this script sourced instead of executed? The end result may be the same in this case, but it looks like an unintentional bug.

Comment (jforissier) [reply common.mk:319]: And there is a missing TAB. Have you tested this?
```
# make buildroot OPTEE_RUST_ENABLE=y
common.mk:319: *** missing separator.  Stop.
```

**Comment by jforissier** [comment]
> > Now a couple more issues/questions...
> > 
> > * I got an error the second time I ran `make -j10 OPTEE_RUST_ENABLE=y CFG_TEE_RAM_VA_SIZE=0x00300000 run`:
> >   ```
> >   mv: '/home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_rust_examples_ext-1.0/examples/supp_plugin-rs/plugin/target/aarch64-unknown-linux-gnu/release/libsyslog_plugin.so' and '/home/jerome/work/optee_repo_qemu_v8/out-br/build/optee_rust_examples_ext-1.0/examples/supp_plugin-rs/plugin/target/aarch64-unknown-linux-gnu/release/ef620757-fa2b-4f19-a1c4-6e51cfe4c0f9.plugin.so' are the same file
> >   ```
> 
> Thank you for your comments. This problem will be resolved in next commit in teaclave-trustzone-sdk.

Cool :+1: 
 
> > * It looks like there is a version mismatch somewhere:
> >   ```
> >   Welcome to Buildroot, type root or test to login
> >   buildroot login: test
> >   $ random-rs -h
> >   random-rs: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by random-rs)
> >   ```
> 
> Sorry, we can't reproduce this error. We compile OP-TEE and rust examples on Ubuntu 20.04 with libc 2.31, and `random-rs` runs normally. We also run `random-rs` successfully in docker `jforissier/optee_os_ci:qemuv8_check`. Could you please give more details?

Indeed, random-rs runs fine in the CI docker image. Something must be wrong in my environment. I will investigate.

```
Welcome to Buildroot, type root or test to login
buildroot login: test
$ random-rs
Invoking TA to generate random UUID...
Invoking done!
Generate random UUID: 823cbc4-ec88-c871-0aee6d28f78f0237
Success
$ 
```

File: common.mk (hunk 6 above)
Comment (fanrong1992) [reply common.mk]: > Why is this script sourced instead of executed? The end result may be the same in this case, but it looks like an unintentional bug.

Yes, execution is better. We found `source $(EDK2_PATH)/edksetup.sh` on line 414 in common.mk, and want to maintain a unified style. Now we change `source` to `./`.

Comment (fanrong1992) [reply common.mk]: > And there is a missing TAB. Have you tested this?
> 
> ```
> # make buildroot OPTEE_RUST_ENABLE=y
> common.mk:319: *** missing separator.  Stop.
> ```

Thanks for your correction. Sorry for my mistake. We have tested this in another directory, because my editor is not configured properly, an error occurred when pasting and copying.

Comment (jenswi-linaro) [reply common.mk]: There's a subtle difference with that line. All the lines for `edk2-clean-common` target runs in the same shell.
Do you see it?

Comment (fanrong1992) [reply common.mk]: I get it, thanks.

Comment (jforissier) [reply common.mk]: How about this:
```diff
diff --git a/common.mk b/common.mk
index f72ea86..6d2d6d0 100644
--- a/common.mk
+++ b/common.mk
@@ -34,6 +34,7 @@ OPTEE_OS_PATH			?= $(ROOT)/optee_os
 OPTEE_CLIENT_PATH		?= $(ROOT)/optee_client
 OPTEE_TEST_PATH			?= $(ROOT)/optee_test
 OPTEE_EXAMPLES_PATH		?= $(ROOT)/optee_examples
+OPTEE_RUST_PATH			?= $(ROOT)/optee_rust
 BUILDROOT_TARGET_ROOT		?= $(ROOT)/out-br/target
 
 # default high verbosity. slow uarts shall specify lower if prefered
@@ -241,6 +242,13 @@ cc-option = $(strip $(call _cc-option,$(1),$(2),$(3)))
 .PHONY: all
 all:
 
+OPTEE_RUST_SET_ENV := source ~/.cargo/env
+
+optee-rust:
+ifeq ($(OPTEE_RUST_ENABLE),y)
+	cd $(OPTEE_RUST_PATH) && ./setup.sh
+endif
+
 ################################################################################
 # Build root
 ################################################################################
@@ -275,6 +283,12 @@ BR2_PACKAGE_OPTEE_EXAMPLES_EXT ?= y
 BR2_PACKAGE_OPTEE_EXAMPLES_EXT_CROSS_COMPILE ?= $(CROSS_COMPILE_S_USER)
 BR2_PACKAGE_OPTEE_EXAMPLES_EXT_SDK ?= $(OPTEE_OS_TA_DEV_KIT_DIR)
 BR2_PACKAGE_OPTEE_EXAMPLES_EXT_SITE ?= $(OPTEE_EXAMPLES_PATH)
+OPTEE_RUST_ENABLE ?= n
+ifeq ($(OPTEE_RUST_ENABLE),y)
+BR2_PACKAGE_OPTEE_RUST_EXAMPLES_EXT ?= y
+BR2_PACKAGE_OPTEE_RUST_EXAMPLES_EXT_CROSS_COMPILE ?= $(CROSS_COMPILE_S_USER)
+BR2_PACKAGE_OPTEE_RUST_EXAMPLES_EXT_SITE ?= $(OPTEE_RUST_PATH)
+endif
 # The OPTEE_OS package builds nothing, it just installs files into the
 # root FS when applicable (for example: shared libraries)
 BR2_PACKAGE_OPTEE_OS_EXT ?= y
@@ -307,7 +321,7 @@ append-var = $(call append-var_,$(1),$(2),$(if $(call y-or-n,$($(1))),,$(double-
 append-br2-vars = $(foreach var,$(filter BR2_%,$(.VARIABLES)),$(call append-var,$(var),$(1)))
 
 .PHONY: buildroot
-buildroot: optee-os
+buildroot: optee-rust optee-os
 	@mkdir -p ../out-br
 	@rm -f ../out-br/build/optee_*/.stamp_*
 	@rm -f ../out-br/extra.conf
@@ -325,7 +339,7 @@ buildroot: optee-os
 		$(DEFCONFIG_FTPM) \
 		--br-defconfig out-br/extra.conf \
 		--make-cmd $(MAKE))
-	@$(MAKE) -C ../out-br all
+	@$(OPTEE_RUST_SET_ENV) $(MAKE) -C ../out-br all
 
 .PHONY: buildroot-clean
 buildroot-clean:
```

Comment (DemesneGH) [reply common.mk]: Hi @jforissier 

It seems that `$(OPTEE_RUST_SET_ENV)` will always execute whether `OPTEE_RUST_ENABLE` is set to `y` or not. Doing `./setup.sh` will install Rust and Cargo, so if `OPTEE_RUST_ENABLE=n` the Cargo will not be installed, then  `$HOME/.cargo/env` doesn't exist.

Comment (jforissier) [reply common.mk]: @DemesneGH duh, my mistake, the definition needs to be inside the conditional of course.

Comment (DemesneGH) [reply common.mk]: @jforissier Updated. Thanks!

File: common.mk [hunk 7]
```diff
@@ -503,6 +510,19 @@ optee-os-common:
 optee-os-clean-common:
 	$(MAKE) -C $(OPTEE_OS_PATH) $(OPTEE_OS_COMMON_FLAGS) clean
 
+################################################################################
+# OP-TEE Rust
+################################################################################
+ifeq ($(OPTEE_RUST_ENABLE),y)
+OPTEE_RUST_SET_ENV=source ~/.cargo/env
+endif
```
Comment (jforissier) [inline common.mk:518]: Please use a single `ifeq` block.

File: common.mk [hunk 8]
```diff
@@ -503,6 +510,19 @@ optee-os-common:
 optee-os-clean-common:
 	$(MAKE) -C $(OPTEE_OS_PATH) $(OPTEE_OS_COMMON_FLAGS) clean
 
+################################################################################
+# OP-TEE Rust
+################################################################################
+ifeq ($(OPTEE_RUST_ENABLE),y)
+OPTEE_RUST_SET_ENV=source ~/.cargo/env
```
Comment (jforissier) [inline common.mk:517]: Space before and after `=`

File: br-ext/package/optee_rust_examples_ext/optee_rust_examples_ext.mk [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+export RUST_TARGET_PATH=$(@D)
+export RUST_COMPILER_RT_ROOT=$(RUST_TARGET_PATH)/rust/rust/src/llvm-project/compiler-rt
+export OPTEE_DIR=$(@D)/../../..
+export OPTEE_OS_DIR=$(OPTEE_DIR)/optee_os
+export OPTEE_CLIENT_DIR=$(OPTEE_DIR)/out-br/build/optee_client_ext-1.0
+export OPTEE_CLIENT_INCLUDE=$(OPTEE_CLIENT_DIR)/out/export/usr/include
+export PATH+=:$(OPTEE_DIR)/toolchains/aarch64/bin
+export VENDOR=qemu_v8.mk
+export OPTEE_OS_INCLUDE=$(OPTEE_DIR)/optee_os/out/arm/export-ta_arm64/include
```
Comment (jforissier) [inline br-ext/package/optee_rust_examples_ext/optee_rust_examples_ext.mk:20]: Space before and after `=` for consistency

**Review by jforissier — APPROVED** [review]
LGTM, with my comments addressed:
`Acked-by: Jerome Forissier <jerome@forissier.org>`

I am still investigating the Glibc version error that I got on my laptop but that is certainly a local environment issue and should not block merging this I think.

**Comment by DemesneGH** [comment]
Hi @jbech-linaro ,
It seems that IBART checks have failed, can you help us figure it out? 
Thanks!

**Comment by jbech-linaro** [comment]
> It seems that IBART checks have failed, can you help us figure it out?

Can't see what really happened, but I suspect this is a timeout. I don't think it's caused by this PR. I'll restart the job.
